Add records for user & rubygem event logging

segiddins opened the pull request:
Extracted from https://github.com/rubygems/rubygems.org/pull/4367 to keep the diff smaller

simi commented on app/models/geoip_info.rb:
```diff
@@ -0,0 +1,19 @@
+class GeoipInfo < OIDC::BaseModel
```
Does it make sense to inherit from OIDC base model here? What about to move this "BaseModel" somewhere in generic namespace and rename to keep it clear what's inside?

segiddins replied:
I can move `OIDC::BaseModel` to `app/application_model.rb` ?

simi replied:
Yup, anything outside OIDC will work for me. I'm not sure about the name, since for DB backed models, there is `ApplicationRecord`. `ApplicationModel` seems ok.

martinemde replied:
With this change, I think this is ready to go.

simi commented on app/models/ip_address.rb:
```diff
@@ -0,0 +1,18 @@
+  has_many :user_events, class_name: "Events::UserEvent", dependent: :nullify
+  has_many :rubygem_events, class_name: "Events::RubygemEvent", dependent: :nullify
+
+  def hash_ip_address!
```
What is reason to store hashed IP address?

segiddins replied:
to allow for lookup of IPs from honeycomb, where we only store the hashed address

segiddins commented on app/models/events/user_event.rb:
```diff
@@ -0,0 +1,4 @@
+class Events::UserEvent < ApplicationRecord
+  belongs_to :user, class_name: "::User"
```
it fails when an `Events::User::` namespace gets introduced later

segiddins commented on app/models/events/recordable.rb:
```diff
@@ -0,0 +1,7 @@
+  extend ActiveSupport::Concern
+
+  included do
+    has_many :events, class_name: "Events::#{name}Event", dependent: :destroy, inverse_of: model_name.param_key
```
yes, see https://github.com/rubygems/rubygems.org/blob/f76d884863507f3d316bf3272c73c1f5193d7233/app/models/events/recordable.rb

simi commented on config/initializers/request_ip_address.rb:
```diff
@@ -0,0 +1,31 @@
+
+Gemcutter::PROXY_TOKEN = ENV["RUBYGEMS_PROXY_TOKEN"].presence
+
+ActiveSupport.on_load(:action_dispatch_request) do
```
Any details on this?

segiddins replied:
This is used to lazily populate the geoip_info and user agent info in the `additional` json on events once they eventually get recorded

martinemde reviewed (commented):
Partial review.

martinemde commented on app/models/geoip_info.rb:
Seems like we will capture the information about the location at the time when the event is stored.

segiddins commented:
I extracted GeoipInfo to a separate table and also added test coverage for the code that was in the initializer (and moved it into a module so we can get code coverage on it)

indirect reviewed (approved):
Nice, this looks like a good baseline to me.

martinemde reviewed (approved):
🌐
